src/webapp/app.py

```diff
@@ -0,0 +1,352 @@
+        # request to the metadata endpoint for your tenant upon creation.
+        # Using the http cache feature will optimize that process as that
+        # data is highly cacheable, and ideally would survive for the whole
+        # user's session. This shouldn't be cached across user sessions.
+        http_cache: dict = session.get("msal_http_response_cache", {})
+
+        # Create an MSAL client (using app configuration values) and providing
+        # our token cache.
+        msal_client = msal.ConfidentialClientApplication(
+            app.config.get("CLIENT_ID"),
+            authority=app.config.get("AUTHORITY"),
+            client_credential=app.config.get("CLIENT_CREDENTIAL"),
+            token_cache=token_cache,
+            http_cache=http_cache,
+        )
+
+        # Invoke the acquire_token flow on the MSAL client for the requested
+        # scope and account. This will use the cache to either retrieve an
+        # existing valid token or will use the refresh token in the cache to
+        # fetch a new access token. A complete cache miss will result in an
+        # error and indicates that a former auth code flow exchange didn't
+        # include this scope in the request.
+        result: "dict[str: Any]" = msal_client.acquire_token_silent(
+            scopes=["https://graph.microsoft.com/User.Read"],
+            account=msal_client.get_accounts()[0],
+        )
+
+        # If we had an expired access token and it needed to be refreshed
+        # we'll want to update our session's token cache to reflect the new
+        # access token and refresh token so next invocation will not require
+        # exchanging the refresh token for the access token again.
+        if token_cache.has_state_changed:
+            session["token_cache"] = token_cache.serialize()
+        session["msal_http_response_cache"] = http_cache
+
+        # If 'result' comes back with an error, that means that the user will
+        # need to go through the auth grant flow again, as the tokens in the
+        # cache cannot be used to make this request for some reason. How you
+        # handle this situation will depend on your desired user experience.
+        # Furthermore, in this sample we asked for all app permissions up
+        # front but in reality we could have asked for no specific scopes in
+        # the initial auth code flow exchange and performed on-demand, step-up
+        # authentication when reaching this based on the MSAL token cache
+        # contents.
+
+        # Simple HTTP Get to graph showing the usage of the retrieved access
+        # token
+        response = requests.get(
+            "https://graph.microsoft.com/v1.0/me",
+            headers={"Authorization": f"Bearer {result['access_token']}"},
+        ).json()
+
+        # Show the "Graph" view
+        return render_template("authenticated/graph.html", graphCallResponse=response)
+
+    @app.get("/admin")
```

Comment:
any desired to move `/admin` and `/graph` right below `/` for code readability purposes?

Reply:
Good call, will move them.